component/topsql/store/store.go

```diff
@@ -165,35 +172,97 @@ func topSQLProtoToMetric(
 }
 
 // transform resource_usage_agent.CPUTimeRecord to util.Metric
-func rsMeteringProtoToMetric(
+func rsMeteringProtoToMetrics(
 	instance, instance_type string,
 	record *rsmetering.ResourceUsageRecord,
-) (m Metric, err error) {
+) (ms []Metric, err error) {
 	tag := tipb.ResourceGroupTag{}
-
-	m.Metric.Name = "cpu_time"
-	m.Metric.Instance = instance
-	m.Metric.InstanceType = instance_type
-
 	tag.Reset()
 	if err = tag.Unmarshal(record.ResourceGroupTag); err != nil {
 		return
 	}
-
-	m.Metric.SQLDigest = hex.EncodeToString(tag.SqlDigest)
-	m.Metric.PlanDigest = hex.EncodeToString(tag.PlanDigest)
+	sqlDigest := hex.EncodeToString(tag.SqlDigest)
+	planDigest := hex.EncodeToString(tag.PlanDigest)
+
+	mCpu := Metric{
+		Metric: topSQLTags{
+			Name:         "cpu_time",
+			Instance:     instance,
+			InstanceType: instance_type,
+			SQLDigest:    sqlDigest,
+			PlanDigest:   planDigest,
+		},
+	}
+	mReadRow := Metric{
+		Metric: topSQLTags{
+			Name:         "read_row",
+			Instance:     instance,
+			InstanceType: instance_type,
+			SQLDigest:    sqlDigest,
+			PlanDigest:   planDigest,
+		},
+	}
+	mReadIndex := Metric{
+		Metric: topSQLTags{
+			Name:         "read_index",
+			Instance:     instance,
+			InstanceType: instance_type,
+			SQLDigest:    sqlDigest,
+			PlanDigest:   planDigest,
+		},
+	}
+	mWriteRow := Metric{
+		Metric: topSQLTags{
+			Name:         "write_row",
+			Instance:     instance,
+			InstanceType: instance_type,
+			SQLDigest:    sqlDigest,
+			PlanDigest:   planDigest,
+		},
+	}
+	mWriteIndex := Metric{
+		Metric: topSQLTags{
+			Name:         "write_index",
+			Instance:     instance,
+			InstanceType: instance_type,
+			SQLDigest:    sqlDigest,
+			PlanDigest:   planDigest,
+		},
+	}
 
 	for i := range record.RecordListCpuTimeMs {
 		tsMillis := record.RecordListTimestampSec[i] * 1000
 		cpuTime := record.RecordListCpuTimeMs[i]
 
-		m.Timestamps = append(m.Timestamps, tsMillis)
-		m.Values = append(m.Values, cpuTime)
+		mCpu.Timestamps = append(mCpu.Timestamps, tsMillis)
+		mCpu.Values = append(mCpu.Values, cpuTime)
+
+		appendMetricRowIndex(i, tsMillis, record.RecordListReadKeys, &mReadRow, &mReadIndex, tag.Label)
+		appendMetricRowIndex(i, tsMillis, record.RecordListWriteKeys, &mWriteRow, &mWriteIndex, tag.Label)
```

Comment:
IS the length of `record.RecordListTimestampSec` equal to `record.RecordListReadKeys`?

Reply:
I think they must be equal, because we use two lists to "simulate" the hashmap, which is guaranteed by TiKV: [L124-L33](https://github.com/tikv/tikv/pull/11085/files#diff-ef1b96ca378cf47b8c626f999c9dcb3bdd2b26c8746340fa9f59b634b40232d2R124-R133)